opensearch/src/test/java/org/opensearch/sql/opensearch/data/type/OpenSearchDataTypeTest.java

```diff
@@ -6,40 +6,323 @@
+        () -> assertEquals(coreType.toString(), type.typeName()),
+        () -> assertEquals(coreType.toString(), type.legacyTypeName()),
+        () -> assertEquals(coreType, type.getExprType())
+    );
+  }
+
+  @ParameterizedTest(name = "{0}")
+  @EnumSource(ExprCoreType.class)
+  public void ofOpenSearchDataTypeFromExprCoreType(ExprCoreType coreType) {
+    var type = OpenSearchDataType.of(coreType);
+    var derivedType = OpenSearchDataType.of(type);
+    assertEquals(type, derivedType);
+  }
+
+  @ParameterizedTest(name = "{0}")
+  @EnumSource(OpenSearchDataType.MappingType.class)
+  public void ofOpenSearchDataTypeFromMappingType(OpenSearchDataType.MappingType mappingType) {
+    assumeFalse(mappingType == Invalid);
+    var type = OpenSearchDataType.of(mappingType);
+    var derivedType = OpenSearchDataType.of(type);
+    assertEquals(type, derivedType);
+  }
+
+  @Test
+  // Test and type added for coverage only
+  public void ofNullMappingType() {
+    assertThrows(IllegalArgumentException.class, () -> OpenSearchDataType.of(Invalid));
+  }
+
+  @Test
+  // cloneEmpty doesn't clone properties, but clones fields and other attributes
+  public void cloneEmpty() {
+    var type = new TestType(OpenSearchDataType.MappingType.Object,
+        Map.of("val", OpenSearchDataType.of(INTEGER)),
+        Map.of("words", OpenSearchDataType.of(STRING)));
+    var clone = type.cloneEmpty();
+
+    assertAll(
+        // can compare because `properties` and `fields` are marked as @EqualsAndHashCode.Exclude
+        () -> assertEquals(type, clone),
+        // read private field `fields`
+        () -> assertEquals(FieldUtils.readField(type, "fields", true),
+            FieldUtils.readField(clone, "fields", true)),
+        () -> assertTrue(clone.getProperties().isEmpty())
+    );
+  }
+
+  // Following structure of nested objects should be flattened
+  // =====================
+  // type
+  //    |- subtype
+  //    |        |
+  //    |        |- subsubtype
+  //    |        |           |
+  //    |        |           |- textWithKeywordType
+  //    |        |           |                    |- keyword
+  //    |        |           |
+  //    |        |           |- INTEGER
+  //    |        |
+  //    |        |- GeoPoint
+  //    |        |- textWithFieldsType
+  //    |                            |- words
+  //    |
+  //    |- text
+  //    |- keyword
+  // =================
+  // as
+  // =================
+  // type : Object
+  // type.subtype : Object
+  // type.subtype.subsubtype : Object
+  // type.subtype.subsubtype.textWithKeywordType : Text
+  //                                           |- keyword : Keyword
+  // type.subtype.subsubtype.INTEGER : INTEGER
+  // type.subtype.geo_point : GeoPoint
+  // type.subtype.textWithFieldsType: Text
+  //                               |- words : Keyword
+  // type.text : Text
+  // type.keyword : Keyword
+  // ==================
+  // Objects are flattened, but Texts aren't
+  // TODO Arrays
```

Comment:
why?  Isn't this part of the nested work?

Reply:
I don't understand the difference at that point between `nested` and `object`. Should be both flattened the same manner?